[Feature](multi-catalog) add refresh for catalog if it needs

q763562998 opened the pull request:
# Proposed changes

Issue Number: close[[#17413](https://github.com/apache/doris/issues/17413)](https://github.com/apache/doris/issues/17413)

## Problem summary

Currently if a catalog is modified externally in doris, doris is not dynamically aware of it.So if a catalog is created with a refresh time configuration, I added a timer for it to refresh the catalog regularly.

## Checklist(Required)

1. Does it affect the original behavior:
   -  Yes
2. Has unit tests been added:
   -  No
3. Has document been added or modified:
   -  Yes
4. Does it need to update dependencies:
   -  No
5. Are there any changes that cannot be rolled back:
   -  No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [[dev@doris.apache.org](mailto:dev@doris.apache.org)](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

q763562998 committed:
add refresh for catalog

q763562998 committed:
add refresh for catalog2

q763562998 committed:
[Feature](multi-catalog)add refresh for catalog if it needs

morningman commented on fe/fe-core/src/main/java/org/apache/doris/PaloFe.java:
```diff
@@ -173,6 +176,12 @@ public static void start(String dorisHomeDir, String pidDir, String[] args, Star
                 qeService.start();
             }
 
+            ScheduledThreadPoolExecutor refreshTimer = ThreadPoolManager.newDaemonScheduledThreadPool(1,
```
This `ScheduledThreadPoolExecutor` should be in `RefreshMgr`.
And there is already a `CatalogMgr` is `Env.java`, why new a `CatalogMgr` here?

Currently, there are 2 main manager:
- RefreshMgr
- CatalogMgr

All refresh method(refresh catalog/db/table/partition) should be in RefreshMgr.
But currently, there is a mistake that "refresh catalog method" is in CatalogMgr.

I think you can do a refactor to move the "refresh catalog method" from CatalogMgr to RefreshMgr.
And this newly added refresh timer thread should be in RefreshMgr too.

morningman replied:
Please change your IDEA formatter to only format the code you modified. Not changing other code line.

q763562998 replied:
1.The reason why I want to create a new `CatalogMgr` is because the `CatalogMgr` needs a timing refresh time, because when the `CatalogMgr` of `Env` is initialized, there is no such timing refresh time, in order to reuse I will subsequently use the `CatalogMgr` of `Env` and assign this refresh time to the` CatalogMg`r of `Env` when it is initialized and constructed
2.Ok,I will  move the "refresh catalog method" from `CatalogMgr` to `RefreshMgr` and putthe refresh timer thread in `RefreshMgr`
3.Ok, I will not change other lines of code

morningman commented on fe/fe-core/src/main/java/org/apache/doris/PaloFe.java:
```diff
@@ -173,6 +176,12 @@ public static void start(String dorisHomeDir, String pidDir, String[] args, Star
 
+            ScheduledThreadPoolExecutor refreshTimer = ThreadPoolManager.newDaemonScheduledThreadPool(1,
+                "refresh-timer-pool", true);
+            Integer refreshTime = 20;
```
What does `20` mean? It is a magic number to me

q763562998 replied:
Indeed it pops up, it is a timer refresh time that will go after every 20 seconds to refresh the catalog that needs to be refreshed

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:
```diff
@@ -88,17 +91,25 @@ public class CatalogMgr implements Writable, GsonPostProcessable {
     // this map will be regenerated from idToCatalog, so not need to persist.
     private final Map<String, CatalogIf> nameToCatalog = Maps.newConcurrentMap();
+
+    private static final Map<String, Integer[]> refreshMap = Maps.newConcurrentMap();
```
Add some comment in code to explain the key value in refreshMap

q763562998 replied:
ok,I will add some comment

q763562998 committed:
[Feature](multi-catalog)move refresh method from CatalogMgr to RefreshManager

q763562998 committed:
Update ProxyMetaStoreClient.java

No modification of the original other files

q763562998 committed:
Update ProxyMetaStoreClient.java

Delete empty lines from the original file

q763562998 committed:
[fix] remove ProxyMetaStoreClient line

q763562998 committed:
Merge remote-tracking branch 'origin/master'

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,26 +19,48 @@
+public class RefreshManager implements Runnable {
     private static final Logger LOG = LogManager.getLogger(RefreshManager.class);
 
+    private static final ScheduledThreadPoolExecutor REFRESH_Timer = ThreadPoolManager.newDaemonScheduledThreadPool(1,
```
```suggestion
    private static final ScheduledThreadPoolExecutor REFRESH_TIMER = ThreadPoolManager.newDaemonScheduledThreadPool(1,
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,26 +19,48 @@
     private static final Logger LOG = LogManager.getLogger(RefreshManager.class);
 
+    private static final ScheduledThreadPoolExecutor REFRESH_Timer = ThreadPoolManager.newDaemonScheduledThreadPool(1,
+            "refresh-timer-pool", true);
```
```suggestion
            "catalog-refresh-timer-pool", true);
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,26 +19,48 @@
+    private static final int REFRESH_TIME = 20;
+    // key is the name of a catalog, value is an array of length 2, used to store
+    // the original refresh time and the current remaining time of the catalog
+    private static Map<String, Integer[]> refreshMap = Maps.newConcurrentMap();
```
```suggestion
    private Map<String, Integer[]> refreshMap = Maps.newConcurrentMap();
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:
```diff
@@ -465,6 +470,20 @@ public CatalogIf replayCreateCatalog(CatalogLog log) throws DdlException {
+                    throw new DdlException("Invalid properties: " + METADATA_REFRESH_INTERVAL_SEC);
+                }
+                Integer[] sec = {metadataRefreshIntervalSec, metadataRefreshIntervalSec};
+                Env.getCurrentEnv().getRefreshManager().getRefreshMap().put(catalogName, sec);
```
wrap the `getRefreshMap().put(catalogName, sec)` into `RefreshManger()`.
Better not expose raw map interface outside

q763562998 replied:
Ok ，I will modify the above code and encapsulation `getRefreshMap().put(catalogName, sec)` into `RefreshManger()`

morningman commented:
Please add FE ut for this.
You can refer to `ColumnPrivTest`.
And there are 2 classes: `TestExternalCatalog` and `TestCatalogProvider`.
These classes are for unit test, you can use `TestCatalogProvider` to add any schema you want to `TestExternalCatalog`,
for mocking some "metadata update" cases

morningman commented:
And also need to rebase to solve the conflict

q763562998 committed:
Merge branch 'master' of https://github.com/apache/doris

 Conflicts:
	fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java

q763562998 committed:
[fix] add FE UT

q763562998 committed:
Merge remote-tracking branch 'origin/master'

q763562998 commented:
> Please add FE ut for this. You can refer to `ColumnPrivTest`. And there are 2 classes: `TestExternalCatalog` and `TestCatalogProvider`. These classes are for unit test, you can use `TestCatalogProvider` to add any schema you want to `TestExternalCatalog`, for mocking some "metadata update" cases

Ok I have added FE ut

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -146,4 +164,36 @@ private void refreshInternalCtlIcebergTable(RefreshTableStmt stmt, Env env) thro
         env.createTable(createTableStmt);
     }
+
+    public void registerIntoRefreshMap(String catalogName, Integer[] sec) {
```
```suggestion
    public void addToRefreshMap(String catalogName, Integer[] sec) {
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:
```diff
@@ -123,6 +124,7 @@ private CatalogIf removeCatalog(long catalogId) {
         if (catalog != null) {
             catalog.onClose();
             nameToCatalog.remove(catalog.getName());
+            Env.getCurrentEnv().getRefreshManager().logOutOfRefreshMap(catalog.getName());
```
Move this to `catalog.onClose()`

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:
```diff
@@ -462,6 +464,20 @@ public CatalogIf replayCreateCatalog(CatalogLog log, boolean isReplay) throws Dd
         try {
             CatalogIf catalog = CatalogFactory.constructorFromLog(log);
+            Map<String, String> props = log.getProps();
+            if (props.containsKey(METADATA_REFRESH_INTERVAL_SEC)) {
```
This should be done after `((ExternalCatalog) catalog).checkProperties()`.
And all properties, including `METADATA_REFRESH_INTERVAL_SEC`, should be checked in `checkProperties()`.
After `checkProperties()`, all properties should be valid.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/datasource/CatalogMgr.java:
```diff
@@ -462,6 +464,20 @@ public CatalogIf replayCreateCatalog(CatalogLog log, boolean isReplay) throws Dd
+                    throw new DdlException("Invalid properties: " + METADATA_REFRESH_INTERVAL_SEC);
+                }
+                Integer[] sec = {metadataRefreshIntervalSec, metadataRefreshIntervalSec};
+                Env.getCurrentEnv().getRefreshManager().registerIntoRefreshMap(catalogName, sec);
```
You also need to modify the method `gsonPostProcess()` in this class.
To add all catalogs which need to refresh to the RefreshManger, or these refresh info will be lost.

You can test this bug with following steps:
1. create a catalog with refresh property.
2. Restart FE, and wait for 2 minutes.
3. Restart FE again, you will see that the catalog won't refresh anymore.

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,26 +19,44 @@
+    private Map<String, Integer[]> refreshMap = Maps.newConcurrentMap();
+
+    public RefreshManager() {
+        REFRESH_TIMER.scheduleAtFixedRate(this::run, 0, REFRESH_TIME, TimeUnit.SECONDS);
```
Should not start the schedule thread in constructor.
Because if do so, every time a `RefreshManager` is created, the schedule thread will start.
But the schedule thread should only start on Master FE.
So you need to add a new method `start()` to start schedule thread.
But this `start()` method should be called in `Env.startMasterOnlyDaemonThreads()`

q763562998 replied:
Ok, I thought `RefreshManager` was a singleton object before such an bug occurred, I will correct it. And add `start()` in `Env.startMasterOnlyDaemonThreads()`

q763562998 committed:
[Feature] remove start of RefreshManager to Env

q763562998 committed:
Merge branch 'master' of https://github.com/apache/doris

q763562998 commented:
Thanks for your suggestions.
- I have moved the `RereshManager.removeFromRefreshMap()` to `catalog.onClose()`.
- In `ExternalCatalog.checkProperties()` I have added a judgment on the refresh parameter METADATA_REFRESH_INTERVAL_SEC. 
- I have  added registration to`RefreshManger` in `CatalogMgr.gsonPostProcess()`.
- And changed the timing mechanism of `RefreshManger` 
- Removed `start()` of `RefreshManager` to `Env`

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,25 +19,38 @@
+    private ScheduledThreadPoolExecutor refreshScheduler = ThreadPoolManager.newDaemonScheduledThreadPool(1,
+            "catalog-refresh-timer-pool", true);
+    // Unit:SECONDS
+    private static final int REFRESH_TIME = 5;
```
```suggestion
    private static final int REFRESH_TIME_SEC = 5;
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -19,25 +19,38 @@
+            "catalog-refresh-timer-pool", true);
+    // Unit:SECONDS
+    private static final int REFRESH_TIME = 5;
+    // key is the name of a catalog, value is an array of length 2, used to store
```
```suggestion
    // key is the id of a catalog, value is an array of length 2, used to store
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -146,4 +159,45 @@ private void refreshInternalCtlIcebergTable(RefreshTableStmt stmt, Env env) thro
+                TimeUnit.SECONDS);
+    }
+
+    private class TaskRefresh implements Runnable {
```
```suggestion
    private class RefreshTask implements Runnable {
```

morningman commented on fe/fe-core/src/main/java/org/apache/doris/catalog/RefreshManager.java:
```diff
@@ -146,4 +159,45 @@ private void refreshInternalCtlIcebergTable(RefreshTableStmt stmt, Env env) thro
+                    timeGroup[1] = current - REFRESH_TIME;
+                    refreshMap.put(catalogId, timeGroup);
+                } else {
+                    String catalogName = Env.getCurrentEnv().getCatalogMgr().getCatalog(catalogId).getName();
```
Catalog may be dropped while you are doing this refresh task.
So need to check if the return value of `getCatalog` is null, or just using `getCatalogOrException` and catch the exception

q763562998 replied:
Ok,I will check if the return value of `getCatalog` is null

q763562998 committed:
[fix] check the value of getCatalog

q763562998 committed:
Merge branch 'master' of https://github.com/apache/doris

q763562998 committed:
[fix] check the style of RefreshManager

morningman commented:
run buildall

morningman reviewed (approved):
LGTM
